fix(docs): fix header indentation causing no page contents sidebar for this page (#2299)
hashicorp/terraform-cdk: website/docs/cdktf/concepts/providers.mdx

  providers in a CDK for Terraform application.
---

# Providers 

> A [provider](/language/providers) is a Terraform plugin that allows users to manage an external API. In your CDK for Terraform (CDKTF) application, you use your preferred programming language to define the [resources](/cdktf/concepts/resources) you want Terraform to manage on one or more providers. 

Reviewer: I wrote that when I first joined...but now that I have a better understanding of how things work, I believe this is more technically correct - lmk what you think!

```suggestion
A [provider](/language/providers) is a plugin that lets Terraform manage an external API. In your CDK for Terraform (CDKTF) application, you use your preferred programming language to define the [resources](/cdktf/concepts/resources) you want Terraform to manage on one or more providers. 
```

> You can install pre-built providers packaged with the required code bindings for your language or add providers to the [`cdktf.json`](/cdktf/create-and-deploy/configuration-file) file and manually generate the required code bindings. To use providers in your application, you can import them from the Terraform Registry or from your local machine.  

Reviewer: ```suggestion
You can install pre-built providers packaged with the required code bindings for your language or add providers to the [`cdktf.json`](/cdktf/create-and-deploy/configuration-file) file and generate code bindings manually. To use providers in your application, you can import them from the Terraform Registry or from your local machine.  
```

> ## What Are Terraform Providers?

Reviewer: Making it shorter so it won't be truncated in the page sidecar!

```suggestion
## What Are Providers?
```

Provider plugins like the [AWS provider](https://registry.terraform.io/providers/hashicorp/aws/latest) or the [cloud-init provider](https://registry.terraform.io/providers/hashicorp/cloudinit/latest/docs) act as a translation layer that allows Terraform to communicate with many different cloud providers, databases, and services.

![diagram: How Terraform uses plugins](images/terraform-plugin-overview.png)

Terraform uses providers to provision [resources](/language/resources), which describe one or more infrastructure objects like virtual networks and compute instances. Each provider on the [Terraform Registry](https://registry.terraform.io/) has documentation detailing available resources and their configuration options.

## Install Pre-Built Providers

> It can take several minutes to generate the code bindings for providers with very large schemas, so we offer several popular providers as pre-built packages. Pre-built providers are a completely optional performance optimization, and you may prefer to generate the code bindings for these providers yourself. For example, you may want to use a different version of that provider than the one in the pre-built package. The [Terraform CDK Providers](https://github.com/orgs/cdktf/repositories?q=cdktf-provider-) page has a complete list, but available pre-built providers include:

Reviewer: This is just a style thing that we changed recently about the way we introduce bulleted lists.

```suggestion
It can take several minutes for CDKTF to generate the code bindings for providers with very large schemas, so we offer several popular providers as pre-built packages. Pre-built providers are a completely optional performance optimization, and you may prefer to generate the code bindings for these providers yourself. For example, you may want to use a different version of that provider than the one in the pre-built package. The [Terraform CDK Providers](https://github.com/orgs/cdktf/repositories?q=cdktf-provider-) page has a complete list, but available pre-built providers include the following options:
```

- [AWS Provider](https://cdk.tf/provider/aws)
- [Google Provider](https://cdk.tf/provider/google)
- [Azure Provider](https://cdk.tf/provider/azurerm)
- [Kubernetes Provider](https://cdk.tf/provider/kubernetes)
- [Docker Provider](https://cdk.tf/provider/docker)
- [Github Provider](https://cdk.tf/provider/github)
- [Null Provider](https://cdk.tf/provider/null)

> These packages are regularly published to NPM / PyPi, and you can treat them as you would any other dependency. The following example shows how to install the AWS provider in TypeScript / Node.

Reviewer: ```suggestion
We regularly publish these packages to NPM / PyPi, and you can treat them like any other dependency. The following example shows how to install the AWS provider in TypeScript / Node.
```

```
npm install @cdktf/provider-aws
```

> When you choose to install a pre-built provider via `npm install`, you should not define that provider again in your `cdktf.json` file. If you receive errors while running `cdktf synth` because of duplicate providers, remove the duplicates from your `cdktf.json` file, delete `tsbuildinfo.json`, and try running `cdktf synth` again.

Reviewer: ```suggestion
When you use `npm install` to install a pre-built provider, you should not define that provider again in your `cdktf.json` file. If you receive errors while running `cdktf synth` because of duplicate providers, remove the duplicates from your `cdktf.json` file, delete `tsbuildinfo.json`, and run `cdktf synth` again.
```

You can also use the [`provider add` command](/cdktf/cli-reference/commands#provider-add) to add providers to your CDKTF application. It will automatically try to install a pre-built provider if available and fall back to generating bindings locally if none was found.

> ## Import Providers

Reviewer: hmm it almost feels like we put this heading in the wrong place! I feel like it should come right above the "You can install pre-built providers....." sentence.

Reviewer: This isn't your fault, Ansgar - it looks like it was already like that, but I'd like to make this opportunistic fix :)

Author: haha, no worries! I'm open for opportunistic fixes!

Author: Okay, got it: https://github.com/hashicorp/terraform-cdk/pull/2299/commits/4f03ceb074c9710a406da5cd90d86c927ce6e5ca
(Is it correct? 😄)

Reviewer: yay!! I like it so much better! Let me go through quickly and just read it, but thank you for doing this!!
